rpcadaptors.go

```diff
@@ -154,21 +215,54 @@ func (cm *rpcConnManager) RemoveByID(id int32) error {
 // This function is safe for concurrent access and is part of the
 // rpcserver.ConnManager interface implementation.
 func (cm *rpcConnManager) RemoveByAddr(addr string) error {
-	replyChan := make(chan error)
-	cm.server.query <- removeNodeMsg{
-		cmp:   func(sp *serverPeer) bool { return sp.Addr() == addr },
-		reply: replyChan,
+	cmp := func(sp *serverPeer) bool { return sp.Addr() == addr }
+	err := cm.removeNode(cmp)
+	if err != nil {
+		netAddr, err := addrStringToNetAddr(addr)
+		if err != nil {
+			return err
+		}
+		return cm.server.connManager.CancelPending(netAddr)
 	}
+	return nil
+}
 
-	// Cancel the connection if it could still be pending.
-	err := <-replyChan
-	if err != nil {
-		cm.server.query <- cancelPendingMsg{
-			addr:  addr,
-			reply: replyChan,
+// disconnectNode disconnects any peers that the provided compare function
+// returns true for.  It applies to both inbound and outbound peers.
+//
+// An error will be returned if no matching peers are found (aka the compare
+// function returns false for all peers).
+//
+// This function is safe for concurrent access.
+func (cm *rpcConnManager) disconnectNode(cmp func(sp *serverPeer) bool) error {
+	state := &cm.server.peerState
+	defer state.Unlock()
+	state.Lock()
+
+	// Check inbound peers.  No callback is passed since there are no additional
+	// actions on disconnect for inbound peers.
+	found := disconnectPeer(state.inboundPeers, cmp, nil)
+	if found {
+		return nil
```

Comment:
It is possible, but most of the time the address will include the port, so it will only match one of them.

That said, I suspect you are correct.  I'd rather not change any existing logic in this regard in this PR though since it's focused on separating the state.  I'll open a separate issue to address this if need be in a different PR.

Reply:
Opened https://github.com/decred/dcrd/issues/3273 per discussion.